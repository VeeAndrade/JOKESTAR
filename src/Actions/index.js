export const updateUser = (username) => ({
  type: 'UPDATE_USER',
  username
});

export const setJoke = (joke) => ({
  type: 'SET_JOKE',
  joke
});

export const favoriteJoke = (joke) => ({
  type: 'FAVORITE_JOKE',
  joke
});

export const addFetchParams = (stateObj) => ({
  type: 'ADD_PARAMS',
  stateObj
});