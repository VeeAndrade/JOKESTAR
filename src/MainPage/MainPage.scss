$faded-purple: rgba(74, 21, 160, 0.452);
$faded-blue: rgba(18, 125, 247, 0.452);
$bright-green: rgb(0, 255, 136);


.users-main-page {
  display: flex; 
  width: 98%;
  height: 42rem;
  margin-top: 2%;
  align-items: center;
  background: linear-gradient($faded-purple, $faded-blue);
}

.user-features-section {
  height: 100%;
  width: 25%;
  border-right: solid 1px $bright-green;
  display: flex;
  flex-direction: column;
}

.feature-button {
  height: 4rem;
  border: solid 1px $bright-green;
  background-color: transparent;
  color: $bright-green;
  margin-top: 2px;
  font-size: 1.4em;

    &:hover {
      background-color: $bright-green;
      color: rebeccapurple;
    }
}

.display-content-section {
  width: 75%;
  height: 100%;
  display: flex;
  align-items: center;
  justify-content: center;
}

.intro {
  color: #fff;
  font-size: 2.5em;
}