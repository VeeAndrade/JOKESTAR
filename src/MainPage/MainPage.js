import React, { Component } from 'react';
import JokeForm from '../JokeForm/JokeForm';
import SafeJoke from '../SafeJoke/SafeJoke'
import NSFWJoke from '../NSFWJoke/NSFWJoke';
import Favorites from '../Favorites/Favorites';
import { getRandomJoke } from '../apiCalls';
import { getSpecificJoke } from '../apiCalls';
import { setJoke } from '../Actions';
import { favoriteJoke } from '../Actions';
import { connect } from 'react-redux';
import './MainPage.scss';

export class MainPage extends Component {
  constructor() {
    super();
    this.state = {
      clickedButton: ''
    }
  }

  setClickedButton = (e) => {
    this.setState({clickedButton: e.target.value})
  }

  determinePageDisplay = () => {
    switch(this.state.clickedButton) {
      case 'nsfw':
        return <JokeForm findClickedBtn={this.setClickedButton}/>
      case 'safejoke':
        return <SafeJoke grabJoke={this.grabRandomJoke} favoriteJoke={this.addJokeToFavorites}/>
      case 'getJoke':
        return <NSFWJoke joke={this.props.joke} getNextJoke={this.grabNSFWJoke} favoriteJoke={this.addJokeToFavorites}/>
      case 'favorites':
        return <Favorites />
      default:
        return <h3 className='intro'>Welcome to JOKESTAR!</h3>
    }
  }

  grabRandomJoke = async () => {
    let promiseJoke = await getRandomJoke()
    this.props.updateJoke(promiseJoke)
    return promiseJoke;
  }

  addJokeToFavorites = (e) => {
    this.props.favoriteJoke(this.props.joke)
  }

  grabNSFWJoke = async () => {
    let promiseJoke = await getSpecificJoke(this.props.fetchParams)
    this.props.updateJoke(promiseJoke)
    return promiseJoke
  }

  render() {
    return (
      <section className='users-main-page'>
        <aside className='user-features-section'>
          <button className='feature-button' value='safejoke' onClick={(event) => this.setClickedButton(event)}>Safe Jokes</button>
          <button className='feature-button' value='nsfw' onClick={(event) => this.setClickedButton(event)}>NSFW</button>
          <button className='feature-button' value='favorites' onClick={(event) => this.setClickedButton(event)}>Favorites</button>
          <button className='feature-button' value='clickMe' onClick={(event) => this.setClickedButton(event)}>Click Me</button>
        </aside>
        <section className='display-content-section'>
          {this.determinePageDisplay()}
        </section>
      </section>
    )
  }
}

export const mapStateToProps = state => ({
  joke: state.joke,
  fetchParams: state.fetchParams
})

export const mapDispatchToProps = dispatch => ({
  updateJoke: joke => dispatch(setJoke(joke)),
  favoriteJoke: joke => dispatch(favoriteJoke(joke)) 
})

export default connect(mapStateToProps, mapDispatchToProps)(MainPage);