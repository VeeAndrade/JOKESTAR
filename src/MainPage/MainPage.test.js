import React from 'react';
import { MainPage, mapStateToProps, mapDispatchToProps } from './MainPage';
import { setJoke } from '../Actions'
import { shallow } from 'enzyme';

describe('MainPage', () => {

  let wrapper

  beforeEach(() => {
    wrapper = shallow(<MainPage
    joke={{id: 42, setup: 'how much does a polarbear weigh?', delivery: 'enough to bresk the ice'}}
    fetchParams={{
      category: 'Any',
      blacklistFlags: [],
      type: []
    }}
    />)
  });

  it('should match the snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });
}); 

describe('mapStateToProps', () => {
  it('should return a joke object', () => {
    const mockState = {
      joke: {
        id: 42,
        setup: 'how much does a polarbear weigh?',
        delivery: 'enough to bresk the ice'
      },
      fetchParams: {
        category: 'Any',
        blacklistFlags: [],
        type: []
      }
    }
    const expected = {
        joke: {
          id: 42,
          setup: 'how much does a polarbear weigh?',
          delivery: 'enough to bresk the ice'
        },
        fetchParams: {
          category: 'Any',
          blacklistFlags: [],
          type: []
        }
    }
    const mappedProps = mapStateToProps(mockState)
    expect(mappedProps).toEqual(expected)
  });
});

describe('mapDispatchToProps', () => {
  it('should call dispatch with setJoke', () => {
    const mockDispatch = jest.fn();
    const mockJoke = {
      id: 42,
      setup: 'how much does a polarbear weigh?',
      delivery: 'enough to bresk the ice'
    }
    const actionToDispatch = setJoke(mockJoke);
    const mappedProps = mapDispatchToProps(mockDispatch);
    mappedProps.updateJoke(mockJoke);
    expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch)
  });
})