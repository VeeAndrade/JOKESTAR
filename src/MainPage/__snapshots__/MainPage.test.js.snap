// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MainPage should match the snapshot 1`] = `
<section
  className="users-main-page"
>
  <aside
    className="user-features-section"
  >
    <button
      className="feature-button"
      onClick={[Function]}
      value="safejoke"
    >
      Safe Jokes
    </button>
    <button
      className="feature-button"
      onClick={[Function]}
      value="nsfw"
    >
      NSFW
    </button>
    <button
      className="feature-button"
      onClick={[Function]}
      value="favorites"
    >
      Favorites
    </button>
    <button
      className="feature-button"
      onClick={[Function]}
      value="clickMe"
    >
      Click Me
    </button>
  </aside>
  <section
    className="display-content-section"
  >
    <h3
      className="intro"
    >
      Welcome to JOKESTAR!
    </h3>
  </section>
</section>
`;
